### ISSUE TITLE
Gradle Error (org.gradle.wrapper.GradleWrapperMain)

### ISSUE THREAD

**Issue by wbrickner** [BrainiumLLC/cargo-mobile#2]
# Issue

Hello,

I have installed all the required dependencies (I believe), and have successfully run an example for desktop, on macOS.

When I now run `cargo android run` I get:

```
Detected connected device: OnePlus 8 Pro (IN2025) with target "aarch64-linux-android"
Error: Could not find or load main class org.gradle.wrapper.GradleWrapperMain
Caused by: java.lang.ClassNotFoundException: org.gradle.wrapper.GradleWrapperMain
error: Failed to assemble APK
    Command "/Users/wbrickner/Documents/Projects/rust-mobile-test/gen/android/testy/gradlew --project-dir
    /Users/wbrickner/Documents/Projects/rust-mobile-test/gen/android/testy assembleArm64Debug --warn" didn't complete
    successfully, exiting with code 1.
```

# Environment

I am running the `Bevy` example with the present commit of this repo, with rust 1.50 nightly.
I believe my NDK version is `21.3.6528147`. 

`$ gradle --version` yields

```
------------------------------------------------------------
Gradle 5.6.2
------------------------------------------------------------

Build time:   2019-09-05 16:13:54 UTC
Revision:     55a5e53d855db8fc7b0e494412fc624051a8e781

Kotlin:       1.3.41
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.14 compiled on March 12 2019
JVM:          11.0.8 (AdoptOpenJDK 11.0.8+10)
OS:           Mac OS X 10.16 x86_64
```

(This gradle output is slightly incorrect in one obvious place, I am actually running macOS Big Sur 11.0.1 w/ Intel ISA).

Hopefully this is enough information to help track down the issue which I assume is internal to the `cargo-mobile` project generation process.

**Comment by francesca64** [comment]
Thanks for reporting this, and for trying out cargo-mobile!

Is `gen/android/testy/gradle/wrapper/gradle-wrapper.jar` present?

Does it work if you run from Android Studio instead?

**Comment by wbrickner** [comment]
Yes the file `gen/android/testy/gradle/wrapper/gradle-wrapper.jar` is present, but it is only 130 bytes large.
This reminds me, when launching on macOS I had a similar issue where the desktop app would hang with a gray screen,
and in the console it reports a PNG decoding error of some sort relating to the image's signature. That file was also exactly 130 bytes.  I would guess there is an issue with how some files are generated / copied into place.

In android studio the build succeeds after Android Studio takes it upon itself to download gradle 5.1.1.  However when I go to run on my device, I get this build output which did not appear before:

```
Android Gradle Plugin: (2 warnings)
   Cannot build selected target ABI: arm64-v8a,armeabi-v7a,armeabi, no suitable splits configured: x86_64;
   Cannot build selected target ABI: arm64-v8a,armeabi-v7a,armeabi, no suitable splits configured: x86;
```

The application _is_ on the device now, but when I launch it, it immediately crashes.

**Comment by wbrickner** [comment]
The corrupted icon.png: ![icon](https://user-images.githubusercontent.com/2095902/100280286-8a953480-2f2d-11eb-8b06-6992ef2af118.png)

```
% hexdump ./assets/branding/icon.png 
0000000 76 65 72 73 69 6f 6e 20 68 74 74 70 73 3a 2f 2f
0000010 67 69 74 2d 6c 66 73 2e 67 69 74 68 75 62 2e 63
0000020 6f 6d 2f 73 70 65 63 2f 76 31 0a 6f 69 64 20 73
0000030 68 61 32 35 36 3a 34 31 39 33 62 61 34 65 63 34
0000040 33 31 38 32 33 38 63 31 62 37 30 33 66 36 61 65
0000050 31 38 39 35 34 37 62 65 36 62 61 35 39 65 61 36
0000060 38 32 61 34 64 63 65 31 34 61 38 36 37 35 36 63
0000070 62 37 38 38 31 64 0a 73 69 7a 65 20 31 35 37 31
0000080 33 0a                                          
0000082
```


The corrupted gradle-wrapper.jar:  [gradle-wrapper.jar.zip](https://github.com/BrainiumLLC/cargo-mobile/files/5599227/gradle-wrapper.jar.zip)
```
% hexdump ./gen/android/testy/gradle/wrapper/gradle-wrapper.jar 
0000000 76 65 72 73 69 6f 6e 20 68 74 74 70 73 3a 2f 2f
0000010 67 69 74 2d 6c 66 73 2e 67 69 74 68 75 62 2e 63
0000020 6f 6d 2f 73 70 65 63 2f 76 31 0a 6f 69 64 20 73
0000030 68 61 32 35 36 3a 33 38 31 64 66 66 38 61 61 34
0000040 33 34 34 39 39 61 61 39 33 62 63 32 35 35 37 32
0000050 62 30 34 39 63 38 63 35 38 36 61 36 37 66 61 66
0000060 66 32 63 30 32 66 33 37 35 65 34 66 32 33 65 31
0000070 37 65 34 39 64 65 0a 73 69 7a 65 20 35 34 33 32
0000080 39 0a                                          
0000082
```

Their contents are identical.

**Comment by francesca64** [comment]
That's weird. Thanks for being so thorough!

If you check `~/.cargo-mobile/templates/platforms/android-studio/gradle/wrapper/gradle-wrapper.jar` and `~/.cargo-mobile/templates/apps/bevy/assets/branding/icon.png`, do those look normal?

**Comment by wbrickner** [comment]
Sorry for the late reply.  Those files also seem to be corrupted, so maybe it's the installation process.

```
% hexdump ~/.cargo-mobile/templates/apps/bevy/assets/branding/icon.png
0000000 76 65 72 73 69 6f 6e 20 68 74 74 70 73 3a 2f 2f
0000010 67 69 74 2d 6c 66 73 2e 67 69 74 68 75 62 2e 63
0000020 6f 6d 2f 73 70 65 63 2f 76 31 0a 6f 69 64 20 73
0000030 68 61 32 35 36 3a 34 31 39 33 62 61 34 65 63 34
0000040 33 31 38 32 33 38 63 31 62 37 30 33 66 36 61 65
0000050 31 38 39 35 34 37 62 65 36 62 61 35 39 65 61 36
0000060 38 32 61 34 64 63 65 31 34 61 38 36 37 35 36 63
0000070 62 37 38 38 31 64 0a 73 69 7a 65 20 31 35 37 31
0000080 33 0a                                          
0000082
```

```
% hexdump ~/.cargo-mobile/templates/platforms/android-studio/gradle/wrapper/gradle-wrapper.jar
0000000 76 65 72 73 69 6f 6e 20 68 74 74 70 73 3a 2f 2f
0000010 67 69 74 2d 6c 66 73 2e 67 69 74 68 75 62 2e 63
0000020 6f 6d 2f 73 70 65 63 2f 76 31 0a 6f 69 64 20 73
0000030 68 61 32 35 36 3a 33 38 31 64 66 66 38 61 61 34
0000040 33 34 34 39 39 61 61 39 33 62 63 32 35 35 37 32
0000050 62 30 34 39 63 38 63 35 38 36 61 36 37 66 61 66
0000060 66 32 63 30 32 66 33 37 35 65 34 66 32 33 65 31
0000070 37 65 34 39 64 65 0a 73 69 7a 65 20 35 34 33 32
0000080 39 0a                                          
0000082
```

**Comment by francesca64** [comment]
No worries! I've found the same sequence of bytes [here](https://github.com/qvacua/vimr/issues/531#issuecomment-637905767), so now it seems pretty obvious that you're seeing Git LFS pointers.

What do you see if you run `git lfs env`?

**Comment by Zeerooth** [comment]
For the record:
I had the exact same issue happening to me (though I was testing on Linux the changes made in [this MR that adds Linux support](https://github.com/BrainiumLLC/cargo-mobile/pull/5)) and gradle-wrapper.jar was corrupt as well, when I ran:
```
cargo mobile init
```
it generated a file with contents like this:
```
version https://git-lfs.github.com/spec/v1
oid sha256:381dff8aa434499aa93bc25572b049c8c586a67faff2c02f375e4f23e17e49de
size 54329
```
So, it looks like git LFS pointer
Running ```git lfs env``` gives the following output:
```
git-lfs/2.9.2 (GitHub; linux amd64; go 1.13.5)
git version 2.25.1

LocalWorkingDir=/home/radek/code/testproj
LocalGitDir=/home/radek/code/testproj/.git
LocalGitStorageDir=/home/radek/code/testproj/.git
LocalMediaDir=/home/radek/code/testproj/.git/lfs/objects
LocalReferenceDirs=
TempDir=/home/radek/code/testproj/.git/lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/home/radek/code/testproj/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file
UploadTransfers=basic,lfs-standalone-file
GIT_ASKPASS=/snap/code/52/usr/share/code/resources/app/extensions/git/dist/askpass.sh
GIT_EXEC_PATH=/usr/lib/git-core
VSCODE_GIT_ASKPASS_MAIN=/snap/code/52/usr/share/code/resources/app/extensions/git/dist/askpass-main.js
VSCODE_GIT_ASKPASS_NODE=/snap/code/52/usr/share/code/code
VSCODE_GIT_IPC_HANDLE=/run/user/1000/vscode-git-0dbedf1070.sock
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

### Temporary workaround

I was able to get the project running again by installing gradle on my system, running ```gradle wrapper``` in my project and then  ```cp -r ./gradle/wrapper/ ./gen/android/{{project name}}/gradle/wrapper/``` 

Of course, it is not optimal, so I'll try to get into the root cause of this problem later

**Comment by wbrickner** [comment]
Hm, this temporary fix does work. 

FYI I installed `git lfs` and then created a new project (with an updated `cargo-mobile`) and encountered the same issue again, not sure if that violates an assumption of your PR that this issue is caused by lack of `git lfs`.

**Comment by wbrickner** [comment]
Jumped the gun (whoops), this temporary fix **does not** work.  I get past the build stage, the application installs on my android device, then it immediately crashes on launch.

**Comment by wbrickner** [comment]
Here's something more useful. When I run `adb logcat | grep -i "Cargo"` and launch the app, I see:

```
01-03 20:39:03.187  1748  3328 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.example.cargo_mobile_test/android.app.NativeActivity bnds=[585,852][856,1247]} from uid 10132 pid 2479
01-03 20:39:03.190   995  1186 E ANDR-PERF-LM: GAMEOPT: getData() 945: isHeavy hint type 1 for app com.example.cargo_mobile_test
01-03 20:39:03.196   995  1186 E ANDR-PERF-LM: GamePerfEngine: runAsync() 463: Hint is for com.example.cargo_mobile_test/android.app.NativeActivity/1 = 0x1003 PID=1748
01-03 20:39:03.197  2479  2479 D BaseDraggingActivity: startActivity bookkeeping intent = Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.example.cargo_mobile_test/android.app.NativeActivity bnds=[585,852][856,1247] }
01-03 20:39:03.201  2479  2479 D Launcher: recentSearchAppAdd# name: com.example.cargo_mobile_test/android.app.NativeActivity#0
01-03 20:39:03.201  2479  2479 D Stats   : Intent.getPackage() is null. therefore try to get package name from componentName.getPackageName(): com.example.cargo_mobile_test
01-03 20:39:03.201  2479  2479 D Stats   : Intent.getPackage() is null. therefore try to get package name from componentName.getPackageName(): com.example.cargo_mobile_test
01-03 20:39:03.202  2479  2479 D ItemClickHandler: onClick app, allappsTransProgress = 0.0, tag = AppInfo(id=-1 type=APP container=-1 screen=-1 cell(-1,-1) span(1,1) minSpan(1,1) rank=0 user=UserHandle{0} title=Cargo Mobile Test hidden=false dirty = false newInstallTag=null componentName=ComponentInfo{com.example.cargo_mobile_test/android.app.NativeActivity} searchResult=RecentSearchAppInfo(id=105 count=10 componentKey=com.example.cargo_mobile_test/android.app.NativeActivity#0 firstLaunch=Sun Jan 03 20:22:20 CST 2021 modified=Sun Jan 03 20:39:03 CST 2021 restore=false) suggestionApp=null)
01-03 20:39:03.207  1748  1863 I ActivityManager: Start proc 9165:com.example.cargo_mobile_test/u0a386 for pre-top-activity {com.example.cargo_mobile_test/android.app.NativeActivity}
01-03 20:39:03.208  1748  1846 D OPFD_CTRL_SVC: [2]handle Starting Window for { com.example.cargo_mobile_test }, Dark? false
01-03 20:39:03.209  1748  1845 D OpRestartProcessManager: updateSelf :  com.example.cargo_mobile_test, size : 30
01-03 20:39:03.225 15511 15560 E ANDR-RAMBOOST_SERVER: ramboost cmd_len = 140 cmd: iop_start -1 com.example.cargo_mobile_test /data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg== false
01-03 20:39:03.229  1748  3381 D OpIrisManager: frontPackageChanged packageName = com.example.cargo_mobile_test activityName = 
01-03 20:39:03.230  1748  3381 D OpSlaNetlinkHelper: frontPackageChanged: com.example.cargo_mobile_test uid:10386 pid:9165 | net.oneplus.launcher uid:10132 pid:2479
01-03 20:39:03.230  1748  3381 D OnePlusDuallteManager: notifyFrontPkgToLlmController :com.example.cargo_mobile_test
01-03 20:39:03.230  1748  3381 D OnePlusDuallteManager: notifyFrontPkgToLlmController setLLMPackageName :com.example.cargo_mobile_test
01-03 20:39:03.231  3286  3660 D OpLowLatencyModeController: setLLMPackageName getFrontPackageName = com.example.cargo_mobile_test
01-03 20:39:03.231  1748  3381 D OpProximityController: noteFrontPackageChanged pkg:com.example.cargo_mobile_test uid:10386 lpkg:net.oneplus.launcher luid:10132
01-03 20:39:03.231  1748  3381 D OpIrisManager: frontActivityChanged packageName = com.example.cargo_mobile_test activityName = com.example.cargo_mobile_test/android.app.NativeActivity
01-03 20:39:03.231  1748  3381 E ScreenModeService: getAppToken ActivityRecord{2e2b2fe u0 com.example.cargo_mobile_test/android.app.NativeActivity t14863}
01-03 20:39:03.231  1748  3381 E ScreenModeService: setRefreshRate token ActivityRecord{2e2b2fe u0 com.example.cargo_mobile_test/android.app.NativeActivity t14863} rate 0 tmpRate 0
01-03 20:39:03.232  1748  3381 D OpQuickReply: setQuickReplyResumed focusedApp ActivityRecord{2e2b2fe u0 com.example.cargo_mobile_test/android.app.NativeActivity t14863} pkgName com.example.cargo_mobile_test
01-03 20:39:03.232  1748  3381 D Hypnus  : hypnusSetScene: uid=10386, pid=9165, name=com.example.cargo_mobile_test, activity=null
01-03 20:39:03.232 15511 15563 E ANDR-RAMBOOST_SERVER: ramboost cmd_len = 142 cmd: iop_start 9165 com.example.cargo_mobile_test /data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg== false
01-03 20:39:03.299  9165  9165 I chatty  : uid=10386(com.example.cargo_mobile_test) identical 39 lines
01-03 20:39:03.308  9165  9165 I chatty  : uid=10386(com.example.cargo_mobile_test) identical 47 lines
01-03 20:39:03.309  9165  9165 I chatty  : uid=10386(com.example.cargo_mobile_test) identical 14 lines
01-03 20:39:03.312  9165  9165 I chatty  : uid=10386(com.example.cargo_mobile_test) identical 1 line
01-03 20:39:03.313  2479  5596 D RecentTasksList: dumpTasks# id=14863, cmp=ComponentInfo{com.example.cargo_mobile_test/android.app.NativeActivity}, uid=0, locked=false, isAppLocked=false
01-03 20:39:03.315  9165  9165 E AndroidRuntime: Process: com.example.cargo_mobile_test, PID: 9165
01-03 20:39:03.315  9165  9165 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.cargo_mobile_test/android.app.NativeActivity}: java.lang.IllegalArgumentException: Unable to find native library cargo_mobile_test using classloader: dalvik.system.PathClassLoader[DexPathList[[],nativeLibraryDirectories=[/data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg==/lib/arm64, /data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64]]]
01-03 20:39:03.315  9165  9165 E AndroidRuntime: Caused by: java.lang.IllegalArgumentException: Unable to find native library cargo_mobile_test using classloader: dalvik.system.PathClassLoader[DexPathList[[],nativeLibraryDirectories=[/data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg==/lib/arm64, /data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64]]]
01-03 20:39:03.317  1748  3381 W ActivityTaskManager:   Force finishing activity com.example.cargo_mobile_test/android.app.NativeActivity
01-03 20:39:03.318  1748 15571 D FrameworkEventCollector: recordAppIssue PN: com.example.cargo_mobile_test
01-03 20:39:03.323   995  1186 E ANDR-PERF-LM: GamePerfEngine: runAsync() 463: Hint is for com.example.cargo_mobile_test/android.app.NativeActivity/1 = 0x1004 PID=1748
01-03 20:39:03.323   995  1186 W ANDR-PERF-LM: MetaMeter: handleAsync() 169: GameOptimization Invalid state for appName com.example.cargo_mobile_test 
01-03 20:39:03.328  1748  1845 D OpRestartProcessManager: Duration is too short, ignore : 114 in com.example.cargo_mobile_test
01-03 20:39:03.351  1748  1843 V WindowManager: Orientation start waiting for draw, mDrawState=DRAW_PENDING in Window{42b29cc u0 Application Error: com.example.cargo_mobile_test}, surfaceController Surface(name=Application Error: com.example.cargo_mobile_test)/@0xd00ec91
01-03 20:39:03.376  1748  1843 D DecorView: onWindowFocusChangedFromViewRoot hasFocus: true, DecorView@c183c85[cargo_mobile_test]
01-03 20:39:03.376  1748  1843 D ViewRootImpl[cargo_mobile_test]: windowFocusChanged hasFocus=true inTouchMode=true
01-03 20:39:03.818  1748  1845 W ActivityTaskManager: Activity top resumed state loss timeout for ActivityRecord{2e2b2fe u0 com.example.cargo_mobile_test/android.app.NativeActivity t14863 f}}
01-03 20:39:03.823  1748  1845 W ActivityTaskManager: Activity pause timeout for ActivityRecord{2e2b2fe u0 com.example.cargo_mobile_test/android.app.NativeActivity t14863 f}}
01-03 20:39:03.825  1748  1845 D OpSlaNetlinkHelper: frontPackageChanged: net.oneplus.launcher uid:10132 pid:2479 | com.example.cargo_mobile_test uid:10386 pid:9165
01-03 20:39:03.826  1748  1845 D OpProximityController: noteFrontPackageChanged pkg:net.oneplus.launcher uid:10132 lpkg:com.example.cargo_mobile_test luid:10386
01-03 20:39:03.929  2479  6997 D RecentTasksList: dumpTasks# id=14863, cmp=ComponentInfo{com.example.cargo_mobile_test/android.app.NativeActivity}, uid=0, locked=false, isAppLocked=false
```

Could this line be a big clue?

```
01-03 20:39:03.315  9165  9165 E AndroidRuntime: Caused by: java.lang.IllegalArgumentException: Unable to find native library cargo_mobile_test using classloader: dalvik.system.PathClassLoader[DexPathList[[],nativeLibraryDirectories=[/data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg==/lib/arm64, /data/app/~~NJeVbUrEklssqGGBlj98Yg==/com.example.cargo_mobile_test-ArH96aPahf26FchWKcfmdg==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64]]]
```

**Comment by Zeerooth** [comment]
@wbrickner That's interesting, it worked for me with the same gradle version, but anyway, this might just be some dumb luck, I think there might still be a conflict with gradle version in the project (it's set to 5.1.1) and gradle-wrapper.jar generated with your system gradle (5.6.2).

What I think is a cleaner solution is (and what I ended up doing in the end) is to just download [gradle-wrapper.jar straight from the project](https://github.com/BrainiumLLC/cargo-mobile/blob/master/templates/platforms/android-studio/gradle/wrapper/gradle-wrapper.jar) and replace your broken gradle-wrapper.jar or better yet! You can run this command to install cargo-mobile instead:

```cargo install --git https://github.com/Zeerooth/cargo-mobile --branch gradle-wrapper-fixes```

I moved gradle-wrapper.jar there from git LFS to the normal repo so there shouldn't be any problems, I hope it will fix them for you. If not, well, we might have some deeper issue present.

**Comment by wbrickner** [comment]
I installed your branch using the exact command you provided, created a blank project using the `winit` example, and I get the same immediate crash and the same `adb logcat` line of interest I highlighted before.

**Comment by wbrickner** [comment]
I have manually removed gradle and created an alias to `gradle 5.1.1`, the issue persists with the same `adb logcat` output:

```
AndroidRuntime: Caused by: java.lang.IllegalArgumentException: Unable to find native library cargo_mobile_test using classloader: dalvik.system.PathClassLoader[DexPathList[[],nativeLibraryDirectories=[/data/app/~~VA5lcmwrAB7Xdo0JgB1gLw==/com.example.cargo_mobile_test-5LZjSL_oXyRSunopLTzKZA==/lib/arm64, /data/app/~~VA5lcmwrAB7Xdo0JgB1gLw==/com.example.cargo_mobile_test-5LZjSL_oXyRSunopLTzKZA==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64]]]
01-04 14:08:11.726  1748  5706 W ActivityTaskManager:   Force finishing activity com.example.cargo_mobile_test/android.app.NativeActivity
```

**Comment by Zeerooth** [comment]
Very peculiar! I was able to replicate this issue (well, in a way) - I created a fresh project with ```cargo mobile init``` and then run ```cargo android run``` and got the same error as you did by looking into adb logcat. But, then I tried to run ```cargo android run``` for the second time and it surprisingly worked! The app launched correctly. 
I was able to replicate it with at least two different project/templates.

It seems like this time this might be some issue connected to the compilation process in rust/rust-mobile, so I'll look into it, but it would be great if you could confirm that the similar thing happens on your machine.

I'll also attach logs from both of my ```cargo android run --verbose``` runs and ```adb logcat | grep -i "Cargo"``` output if anyone wants to take a look at them:
[cargo-android-run-0.txt](https://github.com/BrainiumLLC/cargo-mobile/files/5776530/cargo-android-run-0.txt)
[adb-logcat.txt](https://github.com/BrainiumLLC/cargo-mobile/files/5776531/adb-logcat.txt)
[cargo-android-run-1.txt](https://github.com/BrainiumLLC/cargo-mobile/files/5776534/cargo-android-run-1.txt)

**Comment by wbrickner** [comment]
Yes! It is no longer crashing on launch! Double `cargo android run` works.

Viewing the winit example, I see a black screen, unsure if that is intended behavior.

**Comment by Zeerooth** [comment]
Yeah, it is just a simple window example with a logger connected and nothing else really, by invoking ```cargo run``` to run the project on your PC you can see the logs displayed in the terminal. The logs aren't displayed for Android though (at least for me) but I'm not sure if that's intentional.

Anyway, I'm really glad that it's working for you now! If it's cool with you I think it's worth closing this issue (as MR removing LFS and fixing gradle-wrapper.jar was merged today into master) and creating a new one connected to this need to run ```cargo android run``` twice after creating a new project.

**Comment by wbrickner** [comment]
Perfect, thank you two for all your help!

**Comment by wbrickner** [comment]
@Zeerooth 

After battling with bindgen, I've got rust code running on the Oculus Quest 2, and it feels good.

One problem, I've noticed that when I make a change to my `lib.rs`, and then run `cargo android run`, the old version of my code runs, not the new version.  Then upon second invocation the new version runs.

This sounds very similar to what you experienced, where the first run fails, but the second run includes the `.so` built from the rust side.  I suspect some compilation step lags behind, and we are experiencing the same secondary issue.
